xds/src/main/java/io/grpc/xds/XdsServerWrapper.java

```diff
@@ -328,9 +372,52 @@ public void run() {
             return;
           }
           checkNotNull(update.listener(), "update");
-          filterChains = update.listener().getFilterChains();
+          if (inflight) {
+            releaseSuppliersInFlight();
+          } else {
+            inflight = true;
+          }
+          filterChains.clear();
+          for (FilterChain filterChain: update.listener().getFilterChains()) {
+            filterChains.put(filterChain.getName(), filterChain);
+          }
           defaultFilterChain = update.listener().getDefaultFilterChain();
-          updateSelector();
+          List<FilterChain> allFilterChains = update.listener().getFilterChains();
+          if (defaultFilterChain != null) {
+            allFilterChains = new ArrayList<>(update.listener().getFilterChains());
+            allFilterChains.add(defaultFilterChain);
+          }
+          routingConfigs.clear();
+          List<String> currentRdsResources = new ArrayList<>();
+          for (FilterChain filterChain : allFilterChains) {
+            HttpConnectionManager hcm = filterChain.getHttpConnectionManager();
+            if (hcm.virtualHosts() != null) {
+                routingConfigs.put(filterChain.getName(),
+                        ServerRoutingConfig.create(hcm.httpFilterConfigs(), hcm.virtualHosts()));
+            } else {
+              currentRdsResources.add(hcm.rdsName());
+              RouteDiscoveryState rdsState = routeDiscoveryStates.get(hcm.rdsName());
+              if (rdsState == null) {
+                rdsState = new RouteDiscoveryState(hcm.rdsName(), filterChain,
+                        hcm.httpFilterConfigs());
+                routeDiscoveryStates.put(hcm.rdsName(), rdsState);
+              }
+              xdsClient.watchRdsResource(hcm.rdsName(), rdsState);
```

Comment:
yea its a bug:(, re-register the same resource with the same watcher is not allowed.

Reply:
Actually this re-register was meant to reuse the existing data from ClientXdsClient, calling `.watchRdsResource` would trigger data delivery. 

Did you remember there are particular reasons that we do not allow re-register? Otherwise I may want to change ClientXdsClient to allow this to happen. Otherwise I have to cancel and register and this seems suboptimal.

Reply:
I would rather not change the re-register behavior (maybe we could but we should be careful).
Either I would save the rds response so I can reuse them for the new lds update, or I can cancel and register again and let xdsClient be the source of truth of the newest data. I believe either would work and they eventually will be consistent. The first approach might be a little bit optimized (so little that maybe we can ignore), the second approach is simpler.

Reply:
Cancel+re-watch will produce poor results, as XdsClient will actually throw away the xDS watch and the cached value.

I don't understand the problem. If we don't create a new RouteDiscoveryState, then the existing data is already good. Seems if we delete the `routingConfigs.clear()` line then things would mostly just work.

Reply:
I see. Then we can save the rds update and reuse.
I think we should do `routingConfigs.clear()` because if the filterChain name is duplicate between the previous update and the new update, we can be serving stale data, or even worse, modify filterChain to config mappings.